torchtext/datasets/unsupervised_learning.py

```diff
@@ -0,0 +1,138 @@
+from torchtext.data.functional import custom_replace
+import torch
+from torchtext.utils import download_from_url, extract_archive
+from torchtext.vocab import build_vocab_from_iterator
+from torchtext.data.functional import simple_space_split
+import os
+
+
+_patterns = [(r'<.*>', ''),
+             (r'&amp;', '&'),
+             (r'&lt;', '<'),
+             (r'&gt;', '>'),
+             (r'<ref[^<]*<\/ref>', ''),
+             (r'<[^>]*>', ''),
+             (r'\[http:[^] ]*', '['),
+             (r'\|thumb', ''),
+             (r'\|left', ''),
+             (r'\|right', ''),
+             (r'\|\d+px', ''),
+             (r'\[\[image:[^\[\]]*\|', ''),
+             (r'\[\[category:([^|\]]*)[^]]*\]\]', '[[$1]]'),
+             (r'\[\[[a-z\-]*:[^\]]*\]\]', ''),
+             (r'\[\[[^\|\]]*\|', '[['),
+             (r'\{\{[^\}]*\}\}', ''),
+             (r'\{[^\}]*\}', ''),
+             (r'\[', ''),
+             (r'\]', ''),
+             (r'&[^;]*;', ' '),
+             (r'A', 'a'), (r'B', 'b'), (r'C', 'c'),
+             (r'D', 'd'), (r'E', 'e'), (r'F', 'f'),
+             (r'G', 'g'), (r'H', 'h'), (r'I', 'i'),
+             (r'J', 'j'), (r'K', 'k'), (r'L', 'l'),
+             (r'M', 'm'), (r'N', 'n'), (r'O', 'o'),
+             (r'P', 'p'), (r'Q', 'q'), (r'R', 'r'),
+             (r'S', 's'), (r'T', 't'), (r'U', 'u'),
+             (r'V', 'v'), (r'W', 'w'), (r'X', 'x'),
+             (r'Y', 'y'), (r'Z', 'z'),
+             (r'0', ' zero '), (r'1', ' one '), (r'2', ' two '),
+             (r'3', ' three '), (r'4', ' four '), (r'5', ' five '),
+             (r'6', ' six '), (r'7', ' seven '), (r'8', ' eight '),
+             (r'9', ' nine '),
+             (r'[^a-z\n]+', ' '),
+             (r'\n ', ''),
+             (r'\s+', ' '),
+             (r'\n\s*\n', r'\n')
+             ]
+enwik9_norm_transform = custom_replace(_patterns)
+
+
+def generate_offsets(filename):
+    offsets = []
+    with open(filename) as f:
+        offsets.append(f.tell())
+        while f.readline():
+            offsets.append(f.tell())
+    return offsets
+
+
+def read_lines_from_iterator(data_path, offsets, begin_line, num_lines):
+    with open(data_path) as f:
+        f.seek(offsets[begin_line])
+        for i in range(num_lines):
+            yield f.readline()
+
+
+def normalized_raw_enwik9(input_filename, output_filename):
```

Comment:
nit: maybe it would be a good idea to define what normalize mean in this case?

Reply:
I don't think this "read -> write" function is necessary. It's a simple operation

```
with open(input_filename, 'r') as f1, open(output_filename, 'w') as f2:
    for line in f1:
        line = line.strip() # Removes spaces to the left and right
        if len(line) > 0:
            f2.write(list(enwik9_norm_transform([line]))[0].lstrip())
````

This whole indexing, stripping etc. is strange. Is that really necessary?

Reply:
It's pretty time consuming to normalize the text. The idea is to re-use the normalized text file in the future. Of course, for a smaller text file, we don't need to dump the normalized text into a file and streaming will be a better option.